c10/core/thread_pool.cpp

```diff
@@ -114,21 +114,22 @@ void ThreadPool::main_loop(std::size_t index) {
         LOG(ERROR) << "Exception in thread pool task: unknown";
       }
 
-      // Update status of empty, maybe
-      // Need to recover the lock first
-      lock.lock();
-
-      // Increment count, indicating thread is available.
-      ++available_;
-      if (tasks_.empty() && available_ == total_) {
-        complete_ = true;
-        completed_.notify_one();
-      }
+    } // Destruct tasks before taking the lock
```

Comment:
nit: would be useful to add more details here about the potential deadlock if this is not done.